### PR TITLE
[WIP] Add skeleton function and example

### PR DESCRIPTION
**Pull request by trestletech**
Create a silly example app that has all the right structure.

TODO: 
 - come up with a more compelling Shiny app to test -- but hopefully one that doesn't have additional dependencies
 - Add real testthat unit tests. Right now these are just placeholders.
 - unit tests for the new skeleton function
 - roxygen tests for the new skeleton function
 - more in-line comments sprinkled throughout?
 - a README or pointers to relevant shiny dev center articles?
 - Other skeleton functions offer a way to seed your app with some files. Here it's unclear how we'd direct certain files to app/ui/server and others to R/ helpers. So I just punted on that for now
 - Make return value of this function useful or invisible

### CODE REVIEW

File: R/app_template.R [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+#'     ├── shinytest
+#'     │   └── mytest.R
+#'     ├── testthat.R
+#'     └── testthat
+#'         ├── helper-load.R
+#'         └── test-utils.R
+#' ```
+#'
+#' Some notes about these files:
+#' * app.R is the main application file.
```
Comment (wch) [inline R/app_template.R:33]: Use backticks (and bold?) around file names

File: R/app_template.R [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+#' @param examples Either one of "default", "ask", "all", or any combination of
+#'   "app", "rdir", "module", "shinytest", "testthat", and "server". In an
+#'   interactive session, "default" falls back to "ask"; in a non-interactive
+#'   session, "default" falls back to "all". With "ask", this function will
+#'   prompt the user to select which template items will be added to the new app
+#'   directory. With "all", all template items will be added to the app
+#'   directory.
+#'
+#' @export
+shinyAppTemplate <- function(path = NULL, examples = "default")
```
Comment (wch) [inline R/app_template.R:68]: Should have NULL check for path and throw error.

File: R/app_template.R [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+  choices <- c(
+    app       = "app.R            : Main application file",
+    rdir      = "R/utils.R        : Helper file with R code",
+    module    = "R/my-module.R    : Example module",
+    shinytest = "tests/shinytest/ : Tests using shinytest package",
+    testthat  = "tests/testthat/  : Tests using testthat",
+    server    = "tests/server/    : Tests of server and module code"
+  )
+
+  if (length(examples) == 1 && examples == "default") {
```
Comment (wch) [inline R/app_template.R:79]: Simpler to use `identical(examples, "default")`

File: R/app_template.R [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+    if (interactive()) {
+      examples <- "ask"
+    } else {
+      examples <- "all"
+    }
+  }
+
```
Comment (wch) [inline R/app_template.R:86]:
```
examples <- unique(examples)
```

File: R/app_template.R [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+    # Filter out module files, if applicable.
+    if (!with_module) {
+      files <- files[!grepl("module", files)]
+    }
+
+    # Create any subdirectories if needed
+    dirs <- setdiff(unique(dirname(files)), ".")
+    for (dir in dirs) {
+      dir.create(file.path(tests_dir, dir), recursive = TRUE)
```
Comment (wch) [inline R/app_template.R:184]: `showWarnings = FALSE`

File: R/app_template.R [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+  dir_is_empty <- function(path) {
+    files <- list.files(path, all.files = TRUE, no.. = TRUE)
+    # Ignore .DS_Store files, which are sometimes automatically created on macOS
+    files <- setdiff(files, ".DS_Store")
+    return(length(files) != 0)
+  }
+
+  # Helper to resolve paths relative to our example
+  example_path <- function(path) {
+    system.file("examples", "12_counter", path, package = "shiny")
```
Comment (wch) [inline R/app_template.R:125]: Change name to 12_template.

File: inst/examples/12_counter/R/utils.R [hunk 7]
```diff
@@ -0,0 +1,5 @@
+# Given a numeric vector, convert to strings, sort, and convert back to
```
Comment (wch) [inline inst/examples/12_counter/R/utils.R:1]: Rename this file to sort.R

File: inst/examples/12_counter/app.R [hunk 8]
```diff
@@ -0,0 +1,30 @@
+ui <- fluidPage(
+  # ======== Modules ========
+  # mymoduleUI is defined in R/my-module.R
+  mymoduleUI("mymodule1", "Click counter #1"),
+  mymoduleUI("mymodule2", "Click counter #2"),
+  # =========================
+  wellPanel(
+    sliderInput("size", "Data size", min = 5, max = 20, value = 10),
+    div("Lexically sorted sequence:"),
```
Comment (wch) [inline inst/examples/12_counter/app.R:9]: Instead of hard-coding this, add some metadata which has both options, one of which we keep, depending on choices given to `shinyAppTemplate()`

File: inst/examples/12_counter/R/my-module.R [hunk 9]
```diff
@@ -0,0 +1,23 @@
+mymoduleUI <- function(id, label = "Counter") {
```
Comment (wch) [inline inst/examples/12_counter/R/my-module.R:1]: Move these files to inst/template, then generate 12_counter from that template.

File: R/app_template.R [hunk 10]
```diff
@@ -0,0 +1,259 @@
...
+  r_dir <- file.path(path, "R")
+  if ("rdir" %in% examples) {
+    if (!dirExists(r_dir)) {
+      dir.create(r_dir, recursive = TRUE)
+    }
+    file.copy(example_path("R/utils.R"), r_dir, recursive = TRUE)
+  }
+  if ("module" %in% examples) {
+    if (!dirExists(r_dir)) {
+      dir.create(r_dir, recursive = TRUE)
```
Comment (wch) [inline R/app_template.R:241]: `showWarnings = FALSE`, and one above, and remove `if`

File: inst/examples/12_counter/R/my-module.R [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+  ns <- NS(id)
```
Comment (schloerke) [inline inst/examples/12_counter/R/my-module.R:2]: Add a comment here with something like "All UI identifier values must be namespaced using `id`"

File: inst/examples/12_counter/R/my-module.R [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+  tagList(
+    actionButton(ns("button"), label = label),
+    verbatimTextOutput(ns("out"))
+  )
+}
+
+mymoduleServer <- function(id) {
+  moduleServer(
```
Comment (schloerke) [inline inst/examples/12_counter/R/my-module.R:10]: Add a comment here about why `moduleServer` is required

File: inst/examples/12_counter/tests/server/test-mymodule.R [hunk 13]
```diff
@@ -0,0 +1,18 @@
+# Use testthat just for expectations
+library(testthat)
+
+testServer(mymoduleServer, {
```
Comment (schloerke) [inline inst/examples/12_counter/tests/server/test-mymodule.R:4]: `# See ?testServer for more details`

File: inst/examples/12_counter/tests/server/test-mymodule.R [hunk 14]
```diff
@@ -0,0 +1,18 @@
+# Use testthat just for expectations
```
Comment (schloerke) [inline inst/examples/12_counter/tests/server/test-mymodule.R:1]: Is this file required to be in the `./tests/server` folder?

Doesn't it work out fo the box with `./tests/testthat`?

File: inst/examples/12_counter/R/my-module.R (hunk 9 above)
Comment (wch) [reply inst/examples/12_counter/R/my-module.R]: Done, using the `htmlTemplate()` function
